### ISSUE TITLE
Unable to deploy Wrangler via any example docker compose file

### ISSUE THREAD

**Issue by second2falcon** [Receipt-Wrangler/receipt-wrangler-desktop#148]
Crossposting this issue here too for visibility.

TLDR:
Have been trying to get Wrangler to work for almost 2 weeks now, without any luck.
I, for the life of me, can't figure out why I can't get it to run. I have both completely rewritten the compose files and let them be checked by a friend who is much more knowledgable than me. Nothing seems to work.

Issue in the .github repo (as the example compose files are stored there).
Receipt-Wrangler/.github#2

Any help is appreciated.

**Comment by Noah231515** [comment]
@second2falcon Hey thanks for posting. I'm so sorry I didn't respond sooner, I do not check the .github repository regularly since it's just documentation, that's my fault.

Could I see what your compose file looks like?
Judging from the logs, it looks like service names in the compose files may have been changed.

This is a caveat that I probably should document, but if using the built in proxy the service names must be named the same or else the proxy won't pick them up correctly as the error in the gist points out. They should be named the same as they are in the examples.

**Comment by second2falcon** [comment]
@Noah231515 no worries.

In the log file provided in the gist I used the [example compose file](https://github.com/Receipt-Wrangler/.github/blob/main/examples/postgresql/docker-compose.yaml). 
I also tried setting the container name in the docker compose file, however that still gave the role root does not exist error and the API still exited with code 0.

I just added container_name and then the service name as it is in the compose file. Here are the logs for that:
[Log for slightly modified compose file](https://gist.github.com/second2falcon/957e2af95c4e2250baff5451f690bac6)

**Comment by Noah231515** [comment]
@second2falcon I am seeing the same psgql role "root" does not exist error as well. 
However, in my testing environment everything works correctly despite this. I am running the latest version of all the containers.

Have you created the feature config and, and config in the correct directory per the documentation?

**Comment by second2falcon** [comment]
@Noah231515 that indeed did the trick.

Out of curiosity, is it possible (or will in the future) to specify the API location of the "OpenAI" API? I kinda want to experiment with using a local GPT API and using that with Wrangler.

**Comment by Noah231515** [comment]
@second2falcon Yep! This is a good idea and something I have been wanting to do. I will add it to the backlog, it is fairly straightforward to implement.

Are there any local AI applications that you had in mind? I can make sure the initial implementation works with said application(s).

**Comment by second2falcon** [comment]
@Noah231515 
I was thinking about this one [getumbrel/llama-gpt](getumbrel/llama-gpt). They seem to have models in their repo pre made repo for a lot of different hardware combinations.

I am now also running into another issue with Wrangler.
The console spit out these 2 errors.

`api       | 2023/09/25 12:40:35 /go/api/internal/repositories/categories.go:57 ERROR: column "receipt_categories.receipt_id" must appear in the GROUP BY clause or be used in an aggregate function (SQLSTATE 42803)`
`api       | 2023/09/25 12:38:06 /go/api/internal/repositories/group_members.go:53 record not found`

**Comment by Noah231515** [comment]
@second2falcon That one looks awesome. I'll check it out.

Thanks for this, I'll add a bug to the backlog. It looks like this query does not play nice with postgres.

**Comment by second2falcon** [comment]
@Noah231515 out of curiosity, it seems that the OCR is not working when there is no valid OpenAPI key supplied. Is this correct?

Also, for some reason the receipt image preview is tiny as hell for me. Any idea why?
![image](https://github.com/Receipt-Wrangler/receipt-wrangler-desktop/assets/84903084/74e50f78-5bda-46d8-8671-fc09e0d6949b)

**Comment by Noah231515** [comment]
@second2falcon Yes, as of today, OCR works together with AI, not independently of. OCR is used to pull text from an image, and then AI is used to pull out the various pieces of data needed to fill in the data.


The image preview being too small is a known issue, haven't gotten around to cleaning that up properly yet.

**Comment by second2falcon** [comment]
@Noah231515 ok got it.

Btw I just tried using the llama-gpt docker container on my M1 MacBook and it wasn't working. I'll try on a few other devices and update thise comment if it works elsewhere or if it doesn't.

**Comment by MrSaiclops** [comment]
@second2falcon, were you saying that you got llama-gpt working with Receipt Wrangler? Or just in general?

**Comment by Noah231515** [comment]
@MrSaiclops in general, as there is no way to use any other endpoint at the moment, other than OpenAI's.

**Comment by second2falcon** [comment]
@MrSaiclops @Noah231515 is right. First I wasn't able to get it working on my MacBook standalone, but afterwards, it worked like a charm on the CPUs in my server. As Noah mentioned, there is currently no way to define the endpoint of the API used, hence I mentioned it in [one of my earlier comments](https://github.com/Receipt-Wrangler/receipt-wrangler-desktop/issues/148#issuecomment-1733639987).

**Comment by Noah231515** [comment]
@second2falcon There is now support for llama-gpt specifically. I would consider it _very_ experimental. The infrastructure is there to support it, but I could not get it working locally.

I got llama-gpt working on cpu, and with cuda but it was very unstable on windows 11 with wsl2. 
I could not get llama-gpt working on my debian server, however. I am using GPU pass through on that one so it may have something to do with that.

My system specs are:

GPU: RTX 4090
Memory: 32gb 6ghz memory
CPU: AMD R9 7950x

Anytime I would try to send in anything from Receipt Wrangler it would would crash the API. I tried performance tuning, giving llama-gpt more memory. Nothing actually worked. I also tried GPT4ALL, but their standalone Rest API is still in development, and at least on their master branch building it out of the box didn't work for me either after a lot of fiddling around with it.

This is something I will continue to revisit in the future, but specifically with GPT4ALL I tried sending in some prompts generated from wrangler's api, and it didn't handle them very well on the smaller models. Not only that, it was very slow. However, I'm not convinced GPT4ALL was actually using cuda, it really didn't feel like it even with it enabled.

I tried to send these same prompts through the UI in llama-gpt and resulted in crashes. 🤷🤷

On the plus side, it seems between the two, they use pretty much the same API scheme as OpenAI which makes all of this a lot easier. I didn't have very good luck getting things set up, but I'm sure it will improve in the future. 🤞🤞
